File: FreePascal/Dijkstra.pas
program Dijkstra;

uses
  SysUtils, DateUtils, Math;

const
  INF = MaxInt;  // Valor de infinito

type
  TEdge = record
    dest: LongInt;  // Cambiado a LongInt
    weight: LongInt; // Cambiado a LongInt
  end;

  TAdjList = array of array of TEdge;
  TDistanceArray = array of LongInt; // Cambiado a LongInt

procedure AddEdge(var adjList: TAdjList; src, dest, weight: LongInt); // Cambiado a LongInt
begin
  SetLength(adjList[src], Length(adjList[src]) + 1);
  adjList[src][High(adjList[src])].dest := dest;
  adjList[src][High(adjList[src])].weight := weight;
end;

function Dijkstra(adjList: TAdjList; startNode, nodeCount: LongInt): TDistanceArray; // Cambiado a LongInt
var
  distances: TDistanceArray;
  visited: array of Boolean;
  currentNode, newDist: LongInt; // Cambiado a LongInt
  edge: TEdge;
  i, j: LongInt; // Cambiado a LongInt
begin
  SetLength(distances, nodeCount);
  SetLength(visited, nodeCount);
  
  for i := 0 to nodeCount - 1 do
  begin
    distances[i] := INF; // Inicializamos las distancias como infinitas
    visited[i] := False;  // Marcamos todos los nodos como no visitados
  end;

  distances[startNode] := 0; // Distancia al nodo de inicio es 0

  for i := 0 to nodeCount - 1 do
  begin
    // Encontrar el nodo no visitado con la distancia mínima
    currentNode := -1;
    for j := 0 to nodeCount - 1 do
    begin
      if (not visited[j]) and ((currentNode = -1) or (distances[j] < distances[currentNode])) then
        currentNode := j;
    end;

    if (currentNode = -1) or (distances[currentNode] = INF) then
      Break; // Todos los nodos visitados o no hay acceso

    visited[currentNode] := True; // Marcar como visitado

    // Procesar cada arista del nodo actual
    for edge in adjList[currentNode] do
    begin
      newDist := distances[currentNode] + edge.weight;

      // Solo actualizamos si encontramos una distancia menor
      if newDist < distances[edge.dest] then
        distances[edge.dest] := newDist; // Actualizar la distancia
    end;
  end;

  Dijkstra := distances;  // Retornar las distancias calculadas
end;

procedure GenerateGraph(var adjList: TAdjList; numNodes, numEdges: LongInt); // Cambiado a LongInt
var
  src, dest, weight, i: LongInt; // Cambiado a LongInt
begin
  SetLength(adjList, numNodes);

  Randomize;

  // Conectar nodos aleatoriamente para evitar nodos aislados
  for i := 0 to numEdges - 1 do
  begin
    src := Random(numNodes);
    dest := Random(numNodes);
    while dest = src do
      dest := Random(numNodes); // Asegurarse de no conectar el nodo a sí mismo
    weight := Random(100) + 1; // Peso aleatorio entre 1 y 100
    AddEdge(adjList, src, dest, weight);
  end;
end;

var
  adjList: TAdjList;
  distances: TDistanceArray;
  numNodes, numEdges: LongInt; // Cambiado a LongInt
  startTime, endTime: TDateTime;
  elapsedMillis: Int64;
begin
  numNodes := 10000;  // 10,000 nodos
  numEdges := 50000;  // 50,000 aristas

  // Generar el grafo
  GenerateGraph(adjList, numNodes, numEdges);

  // Ejecutar Dijkstra
  startTime := Now;  // Capturamos el tiempo de inicio
  distances := Dijkstra(adjList, 0, numNodes);
  endTime := Now;  // Capturamos el tiempo al finalizar

  elapsedMillis := MilliSecondsBetween(startTime, endTime);

  WriteLn('Distancia mínima al último nodo: ', distances[numNodes - 1]);
  WriteLn('Tiempo de ejecución: ', elapsedMillis, ' ms');
end.

File: GoLang/Dijkstra.go
package main

import (
	"container/heap"
	"fmt"
	"math"
	"math/rand"
	"time"
)

const INF = math.MaxInt32
const MAX_NODES = 10000 // Define un número fijo de nodos

// Estructura para el nodo
type Node struct {
	index    int
	distance int
}

// Cola de prioridad (min-heap)
type PriorityQueue []*Node

func (pq PriorityQueue) Len() int           { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool { return pq[i].distance < pq[j].distance }
func (pq PriorityQueue) Swap(i, j int)      { pq[i], pq[j] = pq[j], pq[i] }

func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*Node))
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	x := old[n-1]
	*pq = old[0 : n-1]
	return x
}

// Estructura del grafo
type Graph struct {
	nodes int
	edges [MAX_NODES][MAX_NODES]int
}

// Crear un nuevo grafo
func NewGraph(n int) *Graph {
	g := &Graph{
		nodes: n,
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			g.edges[i][j] = 0
		}
	}
	return g
}

// Agregar una arista
func (g *Graph) AddEdge(src, dest, weight int) {
	g.edges[src][dest] = weight
}

// Implementación de Dijkstra
func (g *Graph) Dijkstra(start int) []int {
	distances := make([]int, g.nodes)
	for i := range distances {
		distances[i] = INF
	}
	distances[start] = 0

	pq := &PriorityQueue{}
	heap.Push(pq, &Node{index: start, distance: 0})

	for pq.Len() > 0 {
		currentNode := heap.Pop(pq).(*Node)

		for neighbor := 0; neighbor < g.nodes; neighbor++ {
			if g.edges[currentNode.index][neighbor] > 0 {
				newDist := distances[currentNode.index] + g.edges[currentNode.index][neighbor]
				if newDist < distances[neighbor] {
					distances[neighbor] = newDist
					heap.Push(pq, &Node{index: neighbor, distance: newDist})
				}
			}
		}
	}
	return distances
}

// Función principal
func main() {
	numNodes := 10000 // Número de nodos
	numEdges := 50000 // Número de aristas
	g := NewGraph(numNodes)

	rand.Seed(time.Now().UnixNano())
	for i := 0; i < numEdges; i++ {
		src := rand.Intn(numNodes)
		dest := rand.Intn(numNodes)
		weight := rand.Intn(100) + 1 // Peso aleatorio entre 1 y 100
		g.AddEdge(src, dest, weight)
	}

	startTime := time.Now()
	distances := g.Dijkstra(0) // Calcular distancias desde el nodo 0
	duration := time.Since(startTime)

	fmt.Printf("Distancia mínima al nodo %d: %d\n", numNodes-1, distances[numNodes-1])
	fmt.Printf("Tiempo de ejecución de Dijkstra: %v ms\n", duration.Milliseconds())
}
